Add support for musl-based toolchains

bkuhls opened the pull request:
https://www.musl-libc.org/faq.html
`musl is a “libc”, an implementation of the standard library functionality described in the ISO C and POSIX standards, plus common extensions, intended for use on Linux-based systems.`

This PR contains patches downloaded from
https://github.com/voidlinux/void-packages/tree/master/srcpkgs/kodi/patches
which I tried to port to the current master branch. Using a buildroot-compiled
musl toolchain I could compile current Kodi master with these patches.
Unfortunately I am unable to do runtime testing.

Please note that I did not code the patches so I can not judge their technical
quality. Therefore I consider these patches, especially the fileemu patch,
as WIP to start a discussion.

stefansaraev commented:
thanks!

@Karlson2k @notspiff any chance you can look at dlloader/emu stuff when you have time?

wsnipex commented on xbmc/cores/DllLoader/exports/emu_msvcrt.cpp:
```diff
@@ -1633,7 +1631,7 @@ extern "C"
 
     ret = dll_fgetpos64(stream, &tmpPos);
-#if !defined(TARGET_POSIX) || defined(TARGET_DARWIN) || defined(TARGET_FREEBSD) || defined(TARGET_ANDROID)
+#if !defined(__GLIBC__) || defined(TARGET_DARWIN) || defined(TARGET_FREEBSD) || defined(TARGET_ANDROID)
```
does this mean that musl is not posix? Or that the else branch here is actually posix instead?

bkuhls replied:
I hope my following assumptions are correct, my understand of the code is limited.

Musl follows the POSIX standard, the patch I posted tinkers around with the internal representation of different archs. TARGET_POSIX is defined by Kodi configure in m4/xbmc_arch.m4, based on this patchset this will also the case for musl toolchains but the patch author seems to want to use the non-POSIX code path in emu_msvcrt.cpp for whatever reason.

The problem is that Musl does not provide a macro like \__MUSL\__ to identify it:
http://wiki.musl-libc.org/wiki/FAQ#Q:_why_is_there_no_MUSL_macro_.3F
so other ways need to be found to treat Musl differently from glibc/uClibc. What about adding an internal define to m4/xbmc_arch.m4 like we do for FreeBSD, something like -D_MUSL
```
  i*86*-linux-gnu*|i*86*-*-linux-uclibc*|i*86*-*-linux-musl*)
     AC_SUBST(ARCH_DEFINES, "-DTARGET_POSIX -DTARGET_LINUX -D_LINUX -D_MUSL")
```
and patch emu_msvcrt.cpp like this?
```
-#if !defined(TARGET_POSIX) || defined(TARGET_DARWIN) || defined(TARGET_FREEBSD) || defined(TARGET_ANDROID)
+#if !defined(TARGET_POSIX) || defined(TARGET_DARWIN) || defined(TARGET_FREEBSD) || defined(TARGET_ANDROID) || defined(_MUSL)
```

wsnipex replied:
ok, thats what I suspected. Lets use a the define as you suggested, cause I don't like tinkering with our TARGET_POSIX

Karlson2k commented:
@stefansaraev, will do.

Karlson2k commented:
Musl-based void Linux is not very stable yet.
Failed to run properly on VirtualBox and VMware.
Glibc-versions works fine.

bkuhls commented:
How to prevent spamming this PR while being able to mention it in my local patch file?

MartijnKaijser commented:
you just don't mention it al all or call it "PR 9672" or whatever. Have you seen how how times it already spammed here?

bkuhls commented:
@MartijnKaijser: Yes :( Sorry, I did not notice it to its full extend. Just found https\://gist.github.com/alexpeattie/4729247, I will try to use that from now on.

bkuhls committed:
musl: Fix fileemu

Rebased a patch downloaded from
https://github.com/voidlinux/void-packages/blob/master/srcpkgs/kodi/patches/fix-fileemu.patch

Changed "TARGET_POSIX" -> "__GLIBC__" to "defined(_MUSL)"

Signed-off-by: Bernd Kuhls <bernd.kuhls@t-online.de>

bkuhls committed:
musl: Add support for musl-based toolchains

Patch for configure.ac ported from
https://github.com/voidlinux/void-packages/blob/master/srcpkgs/kodi/patches/fix-musl.patch

Signed-off-by: Bernd Kuhls <bernd.kuhls@t-online.de>

bkuhls committed:
musl: cores/DllLoader/ldt_keeper.c: Fix x86 musl targets

Downloaded from
https://github.com/voidlinux/void-packages/blob/master/srcpkgs/kodi/patches/fix-musl-x86.patch

Signed-off-by: Enno Boland <gottox@voidlinux.eu>
Signed-off-by: Bernd Kuhls <bernd.kuhls@t-online.de>

bkuhls committed:
musl: Remove filewrap from utils/posix/PosixInterfaceForCLog

Downloaded from
https://github.com/voidlinux/void-packages/blob/master/srcpkgs/kodi/patches/remove-filewrap.patch

Signed-off-by: Bernd Kuhls <bernd.kuhls@t-online.de>

bkuhls commented:
Closed due to lack of interest.
